posts/saving-and-analyzing-trending-topics-on-twitter-using-aws-athena-lambda-and-cdk.md

```diff
@@ -0,0 +1,198 @@
+
+Once you have the lambda running for few days, you will be able to view the data in few minutes using [AWS Athena](https://aws.amazon.com/athena/). Athena lets you query your data stored on S3 without the burden of setting up a whole database and having batch processes running. You can create a new database using the AWS Glue service using the console and then create the table using the Athena service.
+
+```sql
+-- Create table
+CREATE EXTERNAL TABLE
+IF NOT EXISTS twitter.trend
+(
+  `name` string,
+  `url` string,
+  `promoted_content` string,
+  `query` string,
+  `tweet_volume` bigint
+)
+ROW FORMAT SERDE 'org.openx.data.jsonserde.JsonSerDe'
+WITH SERDEPROPERTIES (
+  'serialization.format' = '1'
+) LOCATION 's3://S3_BUCKET/'
+TBLPROPERTIES
+('has_encrypted_data'='false');
+```
+
+The query above will create the table, the name of the fields are the same as the one from the JSON stored on S3. Don't forget to replace `S3_BUCKET` with the actual bucket containing the files.
+
+## Querying the table
+
+Here is few queries to showcase what can be done with Athena:
+
+### Select latest trends
+
+```sql
+SELECT
+    date_parse(substr("$path", 61), '%Y-%m-%d %H:%i:%s') as date,
+    tweet_volume as volume,
+    name,
+    url,
+    query,
+    promoted_content as promoted
+FROM twitter.trend
+WHERE tweet_volume > 0
+ORDER BY date DESC;
+```
+
+![Diagram](https://raw.githubusercontent.com/falydoor/blog-usa/twitter-trends/images/2020/08/twitter-trends-latest.png)
+
+### Select trends with highest volume
+
+```sql
+SELECT
+    name,
+    MAX(tweet_volume) as volume
+FROM twitter.trend
+GROUP BY name
+ORDER BY MAX(tweet_volume) DESC;
+```
+
+![Diagram](https://raw.githubusercontent.com/falydoor/blog-usa/twitter-trends/images/2020/08/twitter-trends-volume.png)
+
+### Select trends with highest occurrence
+
+```sql
+SELECT *
+FROM (SELECT name, COUNT(name) as count
+    FROM twitter.trend
+    GROUP BY name)
+ORDER BY count desc;
+```
+
+![Diagram](https://raw.githubusercontent.com/falydoor/blog-usa/twitter-trends/images/2020/08/twitter-trends-count.png)
+
+If you want more details about AWS Glue and Athena, I recommend reading [this blog](https://blog.ippon.tech/aws-glue-and-athena-pt1/) that gives more details on how to get started with these two services.
+
+# Conclusion
+
+Creating a project that uses the Twitter API and analyze the results was pretty easy using the available AWS services. CDK really lets you focus on the architecture design instead of wasting time on trivial things like configurations and permissions.
+
+Companies like [Dataminr](https://www.dataminr.com/) are built on Twitter and provide provides real-time information alerts which are really powerful, especially when they are received before major news organizations. Ippon also developed an internal tool that collects tweets: Customer Oriented Sentiment Analysis (COSA), more details can be found on this [blog](https://blog.ippon.tech/five-interns-walk-into-an-office/).
```

Comment:
Companies like [Dataminr](https://www.dataminr.com/) are built on Twitter and provide real-time information alerts which are really powerful, especially when they are received before major news organizations.